jcl/src/openj9.criu/share/classes/org/eclipse/openj9/criu/CRIUSupport.java

```diff
@@ -438,6 +442,49 @@ public CRIUSupport registerRestoreEnvFile(Path envFile) {
 		return this;
 	}
 
+	/**
+	 * Add new JVM options upon restore. The options will be specified in an options
+	 * file with the form specified in https://www.eclipse.org/openj9/docs/xoptionsfile/
+	 *
+	 * TODO: The first iteration will only support -D options. Subsequent iterations will add more functionality.
+	 *
+	 * @param optionsFile The file that contains the new JVM options to be added on restore
+	 *
+	 * @return this
+	 */
+	public CRIUSupport registerRestoreOptionsFile(Path optionsFile) {
+		String optionsFilePath = optionsFile.toAbsolutePath().toString();
+		this.optionsFile = "-Xoptionsfile=" + optionsFilePath; //$NON-NLS-1$
+
+		J9InternalCheckpointHookAPI.registerPostRestoreHook(RESTORE_OPTIONS_FILE_PRIORITY,
+				"Restore options via options file:" + optionsFilePath, () -> { //$NON-NLS-1$
+					try (BufferedReader optionsFileReader = new BufferedReader(new FileReader(optionsFilePath))) {
+						String entry = null;
+						while ((entry = optionsFileReader.readLine()) != null) {
+							entry = entry.trim();
+							if (!entry.isEmpty()) {
+								if (entry.startsWith("-D")) { //$NON-NLS-1$
+									String entrySplit[] = entry.split("=", 2); //$NON-NLS-1$
+									String propertyValue = entrySplit[1];
+									while (propertyValue.endsWith("\\")) { //$NON-NLS-1$
+										/* line continuation */
+										propertyValue = propertyValue.substring(0, propertyValue.length() - 1);
+										propertyValue = propertyValue.trim();
+										String nextLine = optionsFileReader.readLine();
+										propertyValue += nextLine.trim();
```

Comment:
For a malformed file that has a line continuation mark on the last line, this will throw an NPE.  Not sure it's a problem but it might be better to throw a JVMRestoreException in that case.  (And the code does explicit null checks above, so likely worth doing here)

Reply:
I added NPE to the catch block below, this way we also know the cause of the RestoreException